src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs

```diff
@@ -37,13 +37,14 @@ namespace System
         public static ArraySegment<T> Empty { get; } = new ArraySegment<T>(new T[0]);
 #pragma warning restore CA1825, IDE0300
 
+        [field: NotNull]
```

Comment:
This doesn't seem right.

The value very explicitly can be null, such as due to `default` and the `NotNull` is rather more for return values or generics, not for cases like this where you're explictly contradictory (i.e. saying `T[]?`, meaning can be `null`, and `[NotNull]` simultaneously).

Reply:
`[NotNull]` is a postcondition. It doesn’t change the declared type (`T[]?` still means the field may hold null), but instead indicates that the field is guaranteed to be non-null after the constructor completes.

Reply:
I will take some time to think about this, but here is my repro: [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwGIAHKAQwHMBbMgAgHcyoA7bZi2gE2wGczhcMWgGEAggEYAHACYArGloBJACIBRAAwBmdeoCwAKAMABdbSPiAdMuyVmEHhmxgeF4RE4xRzMrgCePXgBuA2NNWlgyTghmP1oHKABXMAxaAGVxWgBvA1pc2gBtADNsGFxOEFoAOQgMSoTcXABdHLzSbAA3MgwhCKiY31o2DHzGgH5aAH0mcl9g/TzaFtyjMPSACiGR2jIASiylhdoAeiOpqBnaAF5tucO8s4vr5nrcW7yAXxD5vJWzcSQ0hlVFRiBgBplaBQYBhArR3ldaMwYPRAWskSjNupGjs3rkDr9zADNo1aABxaHA0G+NZ7S4APlolLBFgeZFmBw5+k+hn0v160Vi8SSKVS0n231ybU63XCMEiAoGxPGrPZEsWat+oo2zGGJN24ruuROKoRZFxhxNTxe5u5+LChLSYqZ4Mh0Nh8KeyMdaK9mOx5rtfyJOq25IwzppVwZzpZ0zZc25QA===)